CodeFormatCore/src/Format/Analyzer/SemicolonAnalyzer.cpp

```diff
@@ -0,0 +1,135 @@
+#include "CodeFormatCore/Format/Analyzer/SemicolonAnalyzer.h"
+#include "CodeFormatCore/Config/LuaStyleEnum.h"
+#include "CodeFormatCore/Format/FormatState.h"
+#include "LuaParser/Lexer/LuaTokenTypeDetail.h"
+
+SemicolonAnalyzer::SemicolonAnalyzer() {
+}
+
+void SemicolonAnalyzer::Analyze(FormatState &f, const LuaSyntaxTree &t) {
+    if (f.GetStyle().end_statement_with_semicolon == EndStmtWithSemicolon::Keep) {
+        return;// No analysis needed
+    }
+
+    for (auto syntaxNode: t.GetSyntaxNodes()) {
+        if (syntaxNode.IsNode(t)) {
+            if (detail::multi_match::StatementMatch(syntaxNode.GetSyntaxKind(t))) {
+                switch (f.GetStyle().end_statement_with_semicolon) {
+                    case EndStmtWithSemicolon::Always: {
+                        if (!EndsWithSemicolon(syntaxNode, t)) {
+                            AddSemicolon(syntaxNode, t);
+                        }
+                        break;
+                    }
+                    case EndStmtWithSemicolon::Never: {
+                        // is on same line as other statement -> needs to go on new line
+                        if (!IsFirstStmtOfLine(syntaxNode, t)) {
```

Comment:
I'd argue "yes", at least in case of "never" - when the code is not supposed to use semicolons for statements, I'm probably not aiming to use empty statements. 
I see your point, but am tempted to say "if you need/want to use empty statements, use this formatting option with care"... - and put a hint about it in the docs (it doesn't affect any code in the default setting anyway)?